### PR TITLE
qt6: add cmake macros

### PR DESCRIPTION
**Pull request by mjvankampen**
Specify library name and version:  **qt/6.0.1**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

Sets a few imported targets (such as qmake and rcc) as well as the Qt version variables. In addition add the files with Qt macros. This way integration with cmake (such as automoc) and other function (such as qt6_add_resources) works.

@ericLemanissier could you have a look? And do you have any clue what `res/archadatadir/bin/ensure_pro_file.cmake` is needed for? Not sure if we should remove it from the package or keep it there (for android builds maybe?).

### CODE REVIEW

**Comment by mjvankampen** [comment]
I'm not sure what I can do about the failing ci, I don't see how it is related to this pr.

**Comment by mjvankampen** [comment]
So it seems that xkbcommon with gcc 10 is missing a GCC version 10 binary.

**Comment by mathbunnyru** [comment]
Let's wait for: https://github.com/conan-io/conan-center-index/pull/4688

After merging it, packages for new compilers will be available.

**Comment by ericLemanissier** [comment]
does it work even if CMAKE_AUTOMOC_MACRO_NAMES  is not defined ?

**Comment by mjvankampen** [comment]
It seems to otherwise the test package wouldn't compile.

**Comment by ericLemanissier** [comment]
mmm, probably thanks to https://github.com/Kitware/CMake/blob/63aa279343df34d34e0a0fd0c92a444e2528bb8a/Modules/CMakeGenericSystem.cmake#L32 ok

**Comment by ericLemanissier** [comment]
`ensure_pro_file.cmake` is in qtbase/coin directory, so it seems to be used only for qt's own CI. You can safely keep it or remove it, it should not really matter

File: recipes/qt/6.x.x/conanfile.py [hunk 1]
```diff
@@ -545,3 +546,38 @@ def package_info(self):
...
+    set_target_properties(${{QT_CMAKE_EXPORT_NAMESPACE}}::moc PROPERTIES IMPORTED_LOCATION ${{CMAKE_CURRENT_LIST_DIR}}/../../../bin/moc{0})
+endif()
+if(NOT TARGET ${{QT_CMAKE_EXPORT_NAMESPACE}}::rcc)
+    add_executable(${{QT_CMAKE_EXPORT_NAMESPACE}}::rcc IMPORTED)
+    set_target_properties(${{QT_CMAKE_EXPORT_NAMESPACE}}::rcc PROPERTIES IMPORTED_LOCATION ${{CMAKE_CURRENT_LIST_DIR}}/../../../bin/rcc{0})
+endif()
+if(NOT TARGET ${{QT_CMAKE_EXPORT_NAMESPACE}}::uic)
+    add_executable(${{QT_CMAKE_EXPORT_NAMESPACE}}::uic IMPORTED)
+    set_target_properties(${{QT_CMAKE_EXPORT_NAMESPACE}}::uic PROPERTIES IMPORTED_LOCATION ${{CMAKE_CURRENT_LIST_DIR}}/../../../bin/uic{0})
+endif()""".format(extension, int(self.version.split(".")[0]), int(self.version.split(".")[1]))
```
Comment (ericLemanissier) [inline recipes/qt/6.x.x/conanfile.py:572]:
```suggestion
        filecontents = "set(QT_CMAKE_EXPORT_NAMESPACE Qt6)\n"
        filecontents += "set(QT_VERSION_MAJOR %d)\n" % int(self.version.split(".")[0])
        filecontents += "set(QT_VERSION_MINOR %d)\n" % int(self.version.split(".")[1])
        filecontents += "set(QT_VERSION_PATCH %d)\n" % int(self.version.split(".")[2])
        targets = ["moc", "rcc", "tracegen", "cmake_automoc_parser", "qlalr", "qmake"]
        targets.extend(["qdbuscpp2xml", "qdbusxml2cpp"])
        if self.options.gui:
            targets.append("qvkgen")
        if self.options.widgets:
            targets.append("uic")
        if self.options.qttools:
            targets.extend(["qhelpgenerator", "qtattributionsscanner", "windeployqt"])
            targest.extend(["lconvert", "lprodump", "lrelease", "lrelease-pro", "lupdate", "lupdate-pro"])
        if self.options.qtshadertools:
            targets.append("qsb")
        if self.options.qtdeclarative:
            targets.extend(["qmltyperegistrar", "qmlcachegen", "qmllint", "qmlimportscanner"])
            targets.extend(["qmlformat", "qml", "qmlprofiler", "qmlpreview", "qmltestrunner"])
        for target in targets:
            filecontents.append("""if(NOT TARGET ${{QT_CMAKE_EXPORT_NAMESPACE}}::{1})
    add_executable(${{QT_CMAKE_EXPORT_NAMESPACE}}::{1} IMPORTED)
    set_target_properties(${{QT_CMAKE_EXPORT_NAMESPACE}}::{1} PROPERTIES IMPORTED_LOCATION ${{CMAKE_CURRENT_LIST_DIR}}/../../../bin/{1}{0})
endif()
""".format(extension, target)
```
This is coded inside the browser, and not tested whatsoever, so thanks for bearing with the typos and dumb errors

File: recipes/qt/6.x.x/test_package/CMakeLists.txt [hunk 2]
```diff
@@ -10,8 +10,11 @@ conan_set_libcxx()
 conan_output_dirs_setup()
 
 find_package(qt REQUIRED CONFIG)
+set(CMAKE_AUTOMOC 1)
+set(SOURCES test_package.cpp greeter.h)
+qt6_add_resources(SOURCES example.qrc)
```
Comment (ericLemanissier) [inline recipes/qt/6.x.x/test_package/CMakeLists.txt:15]:
```suggestion
set(CMAKE_AUTOMOC ON)
set(CMAKE_AUTORCC ON)
```

File: recipes/qt/6.x.x/test_package/CMakeLists.txt [hunk 3]
```diff
@@ -10,8 +10,11 @@ conan_set_libcxx()
...
 
-add_executable(${PROJECT_NAME} test_package.cpp greeter.h moc_greeter.cpp)
+add_executable(${PROJECT_NAME} ${SOURCES})
```
Comment (ericLemanissier) [inline recipes/qt/6.x.x/test_package/CMakeLists.txt:17]:
```suggestion
add_executable(${PROJECT_NAME} test_package.cpp greeter.h example.qrc)
```

File: recipes/qt/6.x.x/test_package/meson.build [hunk 4]
```diff
@@ -1,6 +1,6 @@
 project('test_package', 'cpp')
-qt5 = import('qt5')
-qt5_dep = dependency('qt5', modules: ['Core'])
-moc_files = qt5.preprocess(moc_headers : 'greeter.h')
+qt6 = import('qt6')
+qt6_dep = dependency('qt6', modules: ['Core'])
+moc_files = qt6.preprocess(moc_headers : 'greeter.h', qresources : 'example.qrc')
```
Comment (ericLemanissier) [inline recipes/qt/6.x.x/test_package/meson.build:4]: if meson is properly supported (at least meson 0.57.0 I think), then you should also remove `False and` in the definition of ` def _meson_supported` in `test_package/conanfile.py`

File: recipes/qt/6.x.x/conanfile.py [hunk 5]
```diff
@@ -545,3 +546,42 @@ def package_info(self):
...
+            targets.append("qsb")
+        if self.options.qtdeclarative:
+            targets.extend(["qmltyperegistrar", "qmlcachegen", "qmllint", "qmlimportscanner"])
+            targets.extend(["qmlformat", "qml", "qmlprofiler", "qmlpreview", "qmltestrunner"])
+        for target in targets:
+            filecontents.append("""if(NOT TARGET ${{QT_CMAKE_EXPORT_NAMESPACE}}::{1})
+    add_executable(${{QT_CMAKE_EXPORT_NAMESPACE}}::{1} IMPORTED)
+    set_target_properties(${{QT_CMAKE_EXPORT_NAMESPACE}}::{1} PROPERTIES IMPORTED_LOCATION ${{CMAKE_CURRENT_LIST_DIR}}/../../../bin/{1}{0})
+endif()
+""".format(extension, target)
```
Comment (ericLemanissier) [inline recipes/qt/6.x.x/conanfile.py:576]:
```suggestion
""".format(extension, target))
```
there may be other typos, sorry :(

Comment (mjvankampen) [reply recipes/qt/6.x.x/conanfile.py:576]: No worries will fix. Like this setup much more complete.

File: recipes/qt/6.x.x/test_package/meson.build (hunk 4 above)
Comment (mjvankampen) [reply recipes/qt/6.x.x/test_package/meson.build:4]: I'm not sure about meson. I guess for a different PR :-).

File: recipes/qt/6.x.x/conanfile.py [hunk 6]
```diff
@@ -538,11 +539,11 @@ def package_info(self):
         def _get_corrected_reqs(requires):
             reqs = []
             for r in requires:
-                reqs.append(r if "::" in r else "qt%s" % r)
+                reqs.append(r if "::" in r else "Qt6%s" % r)
             return reqs
 
         def _create_module(module, requires=[]):
-            componentname = "qt%s" % module
+            componentname = "Qt6%s" % module
```
Comment (ericLemanissier) [inline recipes/qt/6.x.x/conanfile.py:546]: out of curiosity, why this change ?

Comment (mjvankampen) [reply recipes/qt/6.x.x/conanfile.py:546]: It makes this piece somewhat more easy (and readable) as now the component names in the conan package align with the names in Qt. Perfectly fine to change it, but to be honest this seemed simpler.

```python
        for m in os.listdir(os.path.join("lib", "cmake")):
            module = os.path.join("lib", "cmake", m, "%sMacros.cmake" % m)
            if os.path.isfile(module):
                self.cpp_info.components[m].build_modules.append(module)
                self.cpp_info.components[m].builddirs.append(os.path.join("lib", "cmake", m))
            else:
                tools.rmdir(os.path.join("lib", "cmake", m))
```

**Review by ericLemanissier — DISMISSED** [review]
Great work !

File: recipes/qt/6.x.x/conanfile.py [hunk 7]
```diff
@@ -659,14 +666,54 @@ def _create_plugin(pluginname, libname, type, requires):
...
+        if self.options.qtshadertools:
+            targets.append("qsb")
+        if self.options.qtdeclarative:
+            targets.extend(["qmltyperegistrar", "qmlcachegen", "qmllint", "qmlimportscanner"])
+            targets.extend(["qmlformat", "qml", "qmlprofiler", "qmlpreview", "qmltestrunner"])
+        for target in targets:
+            filecontents += """\
+if(NOT TARGET ${{QT_CMAKE_EXPORT_NAMESPACE}}::{1})
+    add_executable(${{QT_CMAKE_EXPORT_NAMESPACE}}::{1} IMPORTED)
+    set_target_properties(${{QT_CMAKE_EXPORT_NAMESPACE}}::{1} PROPERTIES IMPORTED_LOCATION ${{CMAKE_CURRENT_LIST_DIR}}/../../../bin/{1}{0})
```
Comment (prince-chrismc) [inline recipes/qt/6.x.x/conanfile.py:706]: Where does this path point to?

File: recipes/qt/6.x.x/conanfile.py [hunk 8]
```diff
@@ -659,14 +666,54 @@ def _create_plugin(pluginname, libname, type, requires):
...
+endif()
+""".format(extension, target)
```
Comment (prince-chrismc) [inline recipes/qt/6.x.x/conanfile.py:708]: My brain 🤯 ... just an editorial

```suggestion
if(NOT TARGET ${{QT_CMAKE_EXPORT_NAMESPACE}}::{0})
    add_executable(${{QT_CMAKE_EXPORT_NAMESPACE}}::{0} IMPORTED)
    set_target_properties(${{QT_CMAKE_EXPORT_NAMESPACE}}::{0} PROPERTIES IMPORTED_LOCATION ${{CMAKE_CURRENT_LIST_DIR}}/../../../bin/{0}{1})
endif()
""".format(target, extension)
```

File: recipes/qt/6.x.x/conanfile.py [hunk 9]
```diff
@@ -659,14 +666,54 @@ def _create_plugin(pluginname, libname, type, requires):
...
+        if self.options.qttools:
+            targets.extend(["qhelpgenerator", "qtattributionsscanner", "windeployqt"])
+            targets.extend(["lconvert", "lprodump", "lrelease", "lrelease-pro", "lupdate", "lupdate-pro"])
+        if self.options.qtshadertools:
+            targets.append("qsb")
+        if self.options.qtdeclarative:
+            targets.extend(["qmltyperegistrar", "qmlcachegen", "qmllint", "qmlimportscanner"])
+            targets.extend(["qmlformat", "qml", "qmlprofiler", "qmlpreview", "qmltestrunner"])
+        for target in targets:
+            filecontents += """\
```
Comment (prince-chrismc) [inline recipes/qt/6.x.x/conanfile.py:703]: maybe some `textwrap.dedent`?

File: recipes/qt/6.x.x/conanfile.py [hunk 10]
```diff
@@ -659,14 +666,54 @@ def _create_plugin(pluginname, libname, type, requires):
...
+if(NOT TARGET ${{QT_CMAKE_EXPORT_NAMESPACE}}::{1})
+    add_executable(${{QT_CMAKE_EXPORT_NAMESPACE}}::{1} IMPORTED)
+    set_target_properties(${{QT_CMAKE_EXPORT_NAMESPACE}}::{1} PROPERTIES IMPORTED_LOCATION ${{CMAKE_CURRENT_LIST_DIR}}/../../../bin/{1}{0})
+endif()
+""".format(extension, target)
+        tools.save(os.path.join("lib", "cmake", "Qt6Core", "extras.cmake"), filecontents)
+
+        self.cpp_info.components["Qt6Core"].builddirs.append(os.path.join("res","archdatadir","bin"))
+        for m in os.listdir(os.path.join("lib", "cmake")):
+            module = os.path.join("lib", "cmake", m, "%sMacros.cmake" % m)
```
Comment (prince-chrismc) [inline recipes/qt/6.x.x/conanfile.py:713]: Is this using the installed files from the local build?

File: recipes/qt/6.x.x/conanfile.py [hunk 11]
```diff
@@ -659,14 +666,54 @@ def _create_plugin(pluginname, libname, type, requires):
...
+            if os.path.isfile(module):
+                self.cpp_info.components[m].build_modules.append(module)
```
Comment (prince-chrismc) [inline recipes/qt/6.x.x/conanfile.py:715]: `build_modules` support generator specific configurations.... 

I think you might want to checkout https://github.com/conan-io/conan-center-index/pull/4795 which was similar to this PR and done by our expert

...which you will be added to the list once this is merged.

**Review by prince-chrismc — COMMENTED** [review]
Strongly recommend taking a look at #4795

File: recipes/qt/6.x.x/conanfile.py (hunk 7 above)
Comment (mjvankampen) [reply recipes/qt/6.x.x/conanfile.py:706]: to an executable in the bin dir of the package dir.

File: recipes/qt/6.x.x/conanfile.py (hunk 8 above)
Comment (mjvankampen) [reply recipes/qt/6.x.x/conanfile.py:708]: Very fair, saw your previous comment and annoyed the hell out of me when I saw it!

File: recipes/qt/6.x.x/conanfile.py (hunk 10 above)
Comment (mjvankampen) [reply recipes/qt/6.x.x/conanfile.py:713]: Correct, I did not find any absolute paths in there so they should be re-useable.

File: recipes/qt/6.x.x/conanfile.py (hunk 11 above)
Comment (mjvankampen) [reply recipes/qt/6.x.x/conanfile.py:715]: Haha not sure about the expert status, copied most of this from @ericLemanissier. As for the generator specific configurations. 

What would the advantage be of adding it to certain generators only? Because the modules are cmake only?

File: recipes/qt/6.x.x/conanfile.py (hunk 9 above)
Comment (mjvankampen) [reply recipes/qt/6.x.x/conanfile.py]: Added some dedent!

File: recipes/qt/6.x.x/conanfile.py (hunk 7 above)
Comment (prince-chrismc) [reply recipes/qt/6.x.x/conanfile.py]: I am not sure how robust this is 🤔

File: recipes/qt/6.x.x/conanfile.py [hunk 12]
```diff
@@ -659,14 +667,54 @@ def _create_plugin(pluginname, libname, type, requires):
...
+        if self.options.qtshadertools:
+            targets.append("qsb")
+        if self.options.qtdeclarative:
+            targets.extend(["qmltyperegistrar", "qmlcachegen", "qmllint", "qmlimportscanner"])
+            targets.extend(["qmlformat", "qml", "qmlprofiler", "qmlpreview", "qmltestrunner"])
+        for target in targets:
+            filecontents += textwrap.dedent("""\
+                if(NOT TARGET ${{QT_CMAKE_EXPORT_NAMESPACE}}::{0})
+                add_executable(${{QT_CMAKE_EXPORT_NAMESPACE}}::{0} IMPORTED)
+                set_target_properties(${{QT_CMAKE_EXPORT_NAMESPACE}}::{0} PROPERTIES IMPORTED_LOCATION ${{CMAKE_CURRENT_LIST_DIR}}/../../../bin/{0}{1})
```
Comment (ericLemanissier) [inline recipes/qt/6.x.x/conanfile.py:707]:
```suggestion
                    add_executable(${{QT_CMAKE_EXPORT_NAMESPACE}}::{0} IMPORTED)
                    set_target_properties(${{QT_CMAKE_EXPORT_NAMESPACE}}::{0} PROPERTIES IMPORTED_LOCATION ${{CMAKE_CURRENT_LIST_DIR}}/../../../bin/{0}{1})
```

File: recipes/qt/6.x.x/conanfile.py (hunk 11 above)
Comment (prince-chrismc) [reply recipes/qt/6.x.x/conanfile.py:716]: Exactly, it would be very weird to have cmake files include for the other generators

**Comment by prince-chrismc** [comment]
FYI

```
ERROR: Conflict in cmake/3.19.1:
    'cmake/3.19.1' requires 'openssl/1.1.1j' while 'qt/6.0.1' requires 'openssl/1.1.1i'.
    To fix this conflict you need to override the package 'openssl' in your root package.
```

**Comment by prince-chrismc** [comment]
There's a bug when pkc-config is not installed, something is ran before the build_requirements are downloaded

```
ERROR: opengl/system: Error in package_info() method, line 80
        self._fill_cppinfo_from_pkgconfig('gl')
while calling '_fill_cppinfo_from_pkgconfig', line 20
        if not pkg_config.provides:
        ConanException: pkg-config command ['pkg-config', '--print-provides', 'gl', '--print-errors'] failed with error: Command 'pkg-config --print-provides gl --print-errors' returned non-zero exit status 127.
/bin/sh: 1: pkg-config: not found
```

**Comment by prince-chrismc** [comment]
Test package fails with a whitespace to blame

 insert evil laugh

```
qt/6.0.1 (test package): Calling build()
qt/6.0.1 (test package): Building with qmake
Cannot find file: /mnt/c/Users/Chris.
Cannot find file: McArthur/Documents/conan-center-index/recipes/qt/6.x.x/test_package.
ERROR: qt/6.0.1 (test package): Error in build() method, line 100
        self._build_with_qmake()
while calling '_build_with_qmake', line 64
        self.run("qmake %s" % " ".join(args), run_environment=True)
        ConanException: Error 2 while executing qmake /mnt/c/Users/Chris McArthur/Documents/conan-center-index/recipes/qt/6.x.x/test_package DESTDIR=bin
```

I edited line 40

```diff
-args = [self.source_folder, "DESTDIR=bin"]
+args = ["'{}'".format(self.source_folder), "DESTDIR=bin"]
```

File: recipes/qt/6.x.x/conanfile.py (hunk 7 above)
Comment (prince-chrismc) [reply recipes/qt/6.x.x/conanfile.py]: Well it works... I think I get how too 💭 

```cmake
find_package(qt REQUIRED)
foreach(_BUILD_MODULE_PATH ${qt_Core_BUILD_MODULES_PATHS})
    message(${_BUILD_MODULE_PATH})
    include(${_BUILD_MODULE_PATH})
endforeach()

get_target_property(IMPLOC ${QT_CMAKE_EXPORT_NAMESPACE}::uic IMPORTED_LOCATION)
message(${IMPLOC})
```

**Comment by mjvankampen** [comment]
> FYI
> 
> ```
> ERROR: Conflict in cmake/3.19.1:
>     'cmake/3.19.1' requires 'openssl/1.1.1j' while 'qt/6.0.1' requires 'openssl/1.1.1i'.
>     To fix this conflict you need to override the package 'openssl' in your root package.
> ```

Fixed in another pr so this one is old.

File: recipes/qt/6.x.x/conanfile.py (hunk 11 above)
Comment (mjvankampen) [reply recipes/qt/6.x.x/conanfile.py:716]: Ok so the cmake, cmake find package and cmake find package multi generators correct?

Comment (prince-chrismc) [reply recipes/qt/6.x.x/conanfile.py:716]: just find and fin_multi

**Comment by prince-chrismc** [comment]
Might have been another PR or I am mistaken

https://github.com/conan-io/conan-center-index/blob/91b4c79464e12cc7bac8a08eae29e94e8b5ba1a2/recipes/qt/6.x.x/conanfile.py#L195

**Comment by mjvankampen** [comment]
> Might have been another PR or I am mistaken
> 
> https://github.com/conan-io/conan-center-index/blob/91b4c79464e12cc7bac8a08eae29e94e8b5ba1a2/recipes/qt/6.x.x/conanfile.py#L195

This has already been fixed in master. Master is not merged in this branch.

File: recipes/qt/6.x.x/conanfile.py [hunk 13]
```diff
@@ -659,14 +667,54 @@ def _create_plugin(pluginname, libname, type, requires):
...
+            targets.extend(["qmltyperegistrar", "qmlcachegen", "qmllint", "qmlimportscanner"])
+            targets.extend(["qmlformat", "qml", "qmlprofiler", "qmlpreview", "qmltestrunner"])
+        for target in targets:
+            filecontents += textwrap.dedent("""\
+                if(NOT TARGET ${{QT_CMAKE_EXPORT_NAMESPACE}}::{0})
+                    add_executable(${{QT_CMAKE_EXPORT_NAMESPACE}}::{0} IMPORTED)
+                    set_target_properties(${{QT_CMAKE_EXPORT_NAMESPACE}}::{0} PROPERTIES IMPORTED_LOCATION ${{CMAKE_CURRENT_LIST_DIR}}/../../../bin/{0}{1})
+                endif()
+                """.format(target, extension))
+        tools.save(os.path.join("lib", "cmake", "Qt6Core", "extras.cmake"), filecontents)
```
Comment (SpaceIm) [inline recipes/qt/6.x.x/conanfile.py:710]: package folder should be immutable after packaginf. Therefore `package_info()` should not create files in package folder.

Comment (SpaceIm) [inline recipes/qt/6.x.x/conanfile.py:710]: package folder must be immutable after packaging. Therefore package_info() should not create files in package folder.
This block must be called in package()

File: recipes/qt/6.x.x/conanfile.py [hunk 14]
```diff
@@ -554,7 +556,7 @@ def _create_module(module, requires=[]):
             self.cpp_info.components[componentname].requires = _get_corrected_reqs(requires)
 
         def _create_plugin(pluginname, libname, type, requires):
-            componentname = "qt%s" % pluginname
+            componentname = "Qt6%s" % pluginname
```
Comment (SpaceIm) [inline recipes/qt/6.x.x/conanfile.py:559]: Why this modification? It's a breaking change because downstream recipes might already rely on these names, it's like changing the name of a recipe.

File: recipes/qt/6.x.x/conanfile.py (hunk 13 above)
Comment (ericLemanissier) [reply recipes/qt/6.x.x/conanfile.py:710]: sorry the mistake originated from me :/

Comment (SpaceIm) [reply recipes/qt/6.x.x/conanfile.py:710]: There is no hook, but it would be nice to have one.

Comment (prince-chrismc) [reply recipes/qt/6.x.x/conanfile.py:710]: I thought this looked funny! https://github.com/conan-io/conan-center-index/pull/4689#discussion_r593535734

The elegance of the solution though is nice, if this was done on the build machine then the paths would be wrong for the consumer.

File: recipes/qt/6.x.x/conanfile.py (hunk 14 above)
Comment (mjvankampen) [reply recipes/qt/6.x.x/conanfile.py:559]: As this is how Qt names its libraries. It avoids some parsing and keeps it consistent, buuuuuuuut fair point will change.

File: recipes/qt/6.x.x/conanfile.py (hunk 13 above)
Comment (mjvankampen) [reply recipes/qt/6.x.x/conanfile.py:710]: Fixed

File: recipes/qt/6.x.x/conanfile.py (hunk 11 above)
Comment (mjvankampen) [reply recipes/qt/6.x.x/conanfile.py]: Fixed

File: recipes/qt/6.x.x/conanfile.py (hunk 14 above)
Comment (mjvankampen) [reply recipes/qt/6.x.x/conanfile.py]: Fixed

**Comment by mjvankampen** [comment]
CI was running for 3 days, I think something went wrong.

File: recipes/qt/6.x.x/conanfile.py [hunk 15]
```diff
@@ -488,33 +489,73 @@ def build(self):
...
+            if not os.path.isfile(module):
+                tools.rmdir(os.path.join(self.package_folder, "lib", "cmake", m))
+
+        extension = ""
+        if self.settings.os == "Windows":
+            extension = ".exe"
+        filecontents = "set(QT_CMAKE_EXPORT_NAMESPACE Qt6)\n"
+        filecontents += "set(QT_VERSION_MAJOR %d)\n" % int(self.version.split(".")[0])
+        filecontents += "set(QT_VERSION_MINOR %d)\n" % int(self.version.split(".")[1])
+        filecontents += "set(QT_VERSION_PATCH %d)\n" % int(self.version.split(".")[2])
```
Comment (prince-chrismc) [inline recipes/qt/6.x.x/conanfile.py:535]: Would this be an improvement?

```suggestion
        ver = tools.Version(self.version)
        filecontents += "set(QT_VERSION_MAJOR %d)\n" % ver.major)
        filecontents += "set(QT_VERSION_MINOR %d)\n" % ver.minor)
        filecontents += "set(QT_VERSION_PATCH %d)\n" % ver.patch)
```

**Comment by prince-chrismc** [comment]
Might be time to call the infra team, just make sure its not blocked

File: recipes/qt/6.x.x/conanfile.py (hunk 15 above)
Comment (mjvankampen) [reply recipes/qt/6.x.x/conanfile.py]: For sure, thanks

**Review by prince-chrismc — DISMISSED** [review]

File: recipes/qt/6.x.x/conanfile.py [hunk 16]
```diff
@@ -488,33 +489,74 @@ def build(self):
...
+                tools.rmdir(os.path.join(self.package_folder, "lib", "cmake", m))
+
+        extension = ""
+        if self.settings.os == "Windows":
+            extension = ".exe"
+        filecontents = "set(QT_CMAKE_EXPORT_NAMESPACE Qt6)\n"
+        ver = tools.Version(self.version)
+        filecontents += "set(QT_VERSION_MAJOR %d)\n" % ver.major
+        filecontents += "set(QT_VERSION_MINOR %d)\n" % ver.minor
+        filecontents += "set(QT_VERSION_PATCH %d)\n" % ver.patch
```
Comment (prince-chrismc) [inline recipes/qt/6.x.x/conanfile.py:536]:
```suggestion
        filecontents += "set(QT_VERSION_MAJOR %s)\n" % ver.major
        filecontents += "set(QT_VERSION_MINOR %s)\n" % ver.minor
        filecontents += "set(QT_VERSION_PATCH %s)\n" % ver.patch
```

**Review by prince-chrismc — APPROVED** [review]

**Review by ericLemanissier — APPROVED** [review]

File: recipes/qt/6.x.x/conanfile.py [hunk 17]
```diff
@@ -488,33 +489,74 @@ def build(self):
...
+            targets.extend(["qmltyperegistrar", "qmlcachegen", "qmllint", "qmlimportscanner"])
+            targets.extend(["qmlformat", "qml", "qmlprofiler", "qmlpreview", "qmltestrunner"])
+        for target in targets:
+            filecontents += textwrap.dedent("""\
+                if(NOT TARGET ${{QT_CMAKE_EXPORT_NAMESPACE}}::{0})
+                    add_executable(${{QT_CMAKE_EXPORT_NAMESPACE}}::{0} IMPORTED)
+                    set_target_properties(${{QT_CMAKE_EXPORT_NAMESPACE}}::{0} PROPERTIES IMPORTED_LOCATION ${{CMAKE_CURRENT_LIST_DIR}}/../../../bin/{0}{1})
+                endif()
+                """.format(target, extension))
+        tools.save(os.path.join(self.package_folder, self._cmake_executables_file), filecontents)
```
Comment (SpaceIm) [inline recipes/qt/6.x.x/conanfile.py:558]: I'm not sure, but maybe the trick with `find_program` in protobuf recipe would also benefit to this recipe:
https://github.com/conan-io/conan-center-index/blob/493cd2694875f398494264bbceb90c1a196a8f3a/recipes/protobuf/all/conanfile.py#L120-L140

Comment (mjvankampen) [reply recipes/qt/6.x.x/conanfile.py:558]: Qt itself doesn't use find_program either. I think we should stick close to the Qt original approach unless we have a clear reason why something else is required. I think cross compilation might be a good one but I am not sure how that works with Qt, I think a lot more would need to happen. @ericLemanissier do you know? I think it might be a bit out of the scope of this PR (which has been open for 26 days already).

Module and bin more robust is a good one I think.

**Review by SSE4 — APPROVED** [review]
